hawkmoth/__init__.py

```diff
@@ -37,6 +37,25 @@ class _AutoBaseDirective(SphinxDirective):
 
     _docstring_types = None
 
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        # Map deprecated options; to be removed.
+        if self.env.config.cautodoc_root:
+            self.env.config.hawkmoth_root = self.env.config.cautodoc_root
+            fmt = 'cautodoc_root option is deprecated, please use hawkmoth_root.'
+            self.logger.warning(fmt, location=(self.env.docname, self.lineno))
+
+        if self.env.config.cautodoc_transformations:
+            fmt = 'cautodoc_transformations option is deprecated, please use hawkmoth_transformations.'  # noqa: E501
+            self.logger.warning(fmt, location=(self.env.docname, self.lineno))
+            self.env.config.hawkmoth_transformations = self.env.config.cautodoc_transformations
+
```

Comment:
The thing that bugs me about this one is that I think I screwed up the whole cautodoc_transformations thing, and it needs to be changed, again. (Read: It bugs me because it rubs my own mistakes in my face.)

The problem is that the configuration is pickled for subsequent Sphinx runs, and function references don't pickle nicely.

I think the solution is to turn transformations to Sphinx events, which does have the nice property of being able to extend Hawkmoth with proper Sphinx extensions that listen to those events.

If we can squeese that into the next release, along with this and the C++ stuff, there aren't going to be so many deprecation and rename steps. If not, it's a nasty experience for the users I'm afraid.

Reply:
Hmm, I haven't really thought about that. I remember liking the changes back then and I may not know enough of Python to really grasp the issues you're referring to.

Do you have an idea of how it should / could be done instead?

Reply:
I liked the changes way back too, but only later realized every no-op rebuild causes:

```
updating environment: [config changed ('hawkmoth_transformations')] 7 added, 0 changed, 0 removed
```

which causes the rewrite of the output, even when nothing changed.

See issue #52.

Reply:
Ah, I get it now. Good point...

So, it's unfortunate if we deprecate this now only to deprecate it again soon when we address that issue. It's also a bit conspicuous to rename every variable except this one. And fixing that issue 1st would just stretch me a bit too thin for what I had in mind.

It may be an easy fix, but honestly I have no idea. Is there an interim solution you'd accept here?

Reply:
I've got some ideas and code for the fix myself, but it might take a while before I get around to polishing it. I guess the question is, do you have a deadline for having a release with the features you need? I don't mind having multiple deprecation steps *between* releases, but I'd really like to get this sorted out before the next release.

Another interim solution might be just not touching this for now, even if the name is inconsistent.

Reply:
> I've got some ideas and code for the fix myself, but it might take a while before I get around to polishing it.

Nice!

> I guess the question is, do you have a deadline for having a release with the features you need? I don't mind having multiple deprecation steps _between_ releases, but I'd really like to get this sorted out before the next release.

Well, it's not a given that we'll take this route. Bottom line, I'd like to have something within a week that competes with breathe, which I'm still optimistic about. Other alternatives would imply moving Sphinx documentation to something else entirely, so I'm placing lower bets on those.

Technically, I think there would be no competition if not for the fear of taking a dependency on an untested solution.

And of course getting the changes upstreamed and an official release in pypi would be important in building that confidence. Not trying to put any pressure on you as my reviewer! :sweat_smile: Just bearing out the facts.

> Another interim solution might be just not touching this for now, even if the name is inconsistent.

I was actually thinking the same. Thought I'd prefer to use the new variables on new projects if we already know we want them. I can go either way honestly...

Reply:
Unrelated note: I forgot to update the docs with the new variables... So this is incomplete anyway.